applications/FluidDynamicsApplication/python_scripts/navier_stokes_two_fluids_hydraulic_solver.py

```diff
@@ -0,0 +1,692 @@
+
+        # Prepare distance correction for next step
+        self._GetDistanceModificationProcess().ExecuteFinalizeSolutionStep()
+
+        # Finalize the solver current step
+        self._GetSolutionStrategy().FinalizeSolutionStep()
+
+        # Acceleration for generalised alpha time integration method.
+        self.__CalculateTimeIntegrationAcceleration()
+
+        # Artificial viscosity for shock capturing.
+        if self.artificial_viscosity:
+            self.__CalculateArtificialViscosity()
+
+
+    def _ComputeInitialStepWaterVolume(self):
+
+        # Here the initial water volume of the system is calculated without considering inlet and outlet flow rate
+        if self.__initial_system_volume is None:
+            self.__initial_system_volume = KratosCFD.FluidAuxiliaryUtilities.CalculateFluidNegativeVolume(
+                self.GetComputingModelPart())
+
+        # Calculate the inlet and outlet volume discharges
+        outlet_discharge = KratosCFD.FluidAuxiliaryUtilities.CalculateFlowRateNegativeSkin(self.GetComputingModelPart(), KratosMultiphysics.OUTLET)
+        inlet_discharge = KratosCFD.FluidAuxiliaryUtilities.CalculateFlowRateNegativeSkin(self.GetComputingModelPart(), KratosMultiphysics.INLET)
+        current_dt = self.main_model_part.ProcessInfo[KratosMultiphysics.DELTA_TIME]
+        inlet_volume = -current_dt * inlet_discharge
+        outlet_volume = current_dt * outlet_discharge
+
+        # System water volume is calculated for current time step considering inlet and outlet discharge.
+        system_volume = inlet_volume + self.__initial_system_volume - outlet_volume
+        self.__initial_system_volume = system_volume
+
+
+
+    def _ComputeVolumeError(self):
+        if self.mass_source:
+            water_volume_after_transport = KratosCFD.FluidAuxiliaryUtilities.CalculateFluidNegativeVolume(self.GetComputingModelPart())
+
+            volume_error = (water_volume_after_transport -  self.__initial_system_volume) /  self.__initial_system_volume
+        else:
+            volume_error=0
+
+        self.main_model_part.ProcessInfo.SetValue(KratosCFD.VOLUME_ERROR, volume_error)
+
+
+
+    def __CalculateTimeIntegrationAcceleration(self):
+        # Acceleration for generalised alpha time integration method.
+        alpha_m=0.5*((3-self.rho_inf)/(1+self.rho_inf))
+        alpha_f=1/(1+self.rho_inf)
+        gamma= 0.5+alpha_m-alpha_f
+        dt=self.main_model_part.ProcessInfo[KratosMultiphysics.DELTA_TIME]
+        for node in self.GetComputingModelPart().Nodes:
+            vn=node.GetSolutionStepValue(KratosMultiphysics.VELOCITY,1)
+            v = node.GetSolutionStepValue(
+                KratosMultiphysics.VELOCITY, 0)
+            acceleration_n=node.GetValue(KratosMultiphysics.ACCELERATION)
+            acceleration_alpha_method=(v-vn)/(gamma*dt)+((gamma-1)/gamma)*acceleration_n
+            node.SetValue(KratosMultiphysics.ACCELERATION,acceleration_alpha_method)
```

Comment:
Not for this PR, but we should consider implementing this in c++.

Reply:
Yes, I agree.